### ISSUE TITLE
Images in markdown are not found by pandocomatic (md to docx conversion)

### ISSUE THREAD

**Issue by agusmba** [htdebeer/pandocomatic#45]
This happens if the mardown source is not in the current directory where you are calling pandocomatic.

I know pandoc doesn't find them either, that's why in my home-baked scripts I had to `cd` back and forth into directories containing the `.md` files before calling pandoc.

However reading the "converting a directory tree of documents" feature of pandocomatic, I expected that it would work, but alas:

```text
$ pandocomatic -b -c ./pandocomatic-45/pandocomatic-45.yaml -i pandocomatic-45 -o out
7 commands to execute to convert 'pandocomatic-45' to 'out'.
(7) + convert pandocomatic-45; enter out
(6) + convert pandocomatic-45/img; enter out/img
(5)   - copy test-img.png
(3)   - convert pandocomatic-45.md -> pandocomatic-45.docx
pandoc  --standalone --to=docx --output=out/pandocomatic-45.docx
[warning] Could not fetch resource './img/test-img.png': PandocResourceNotFound "./img/test-img.png"
[warning] Could not fetch resource 'img/test-img.png': PandocResourceNotFound "img/test-img.png"
[warning] Could not fetch resource './test-img-2.png': PandocResourceNotFound "./test-img-2.png"
[warning] Could not fetch resource 'test-img-2.png': PandocResourceNotFound "test-img-2.png"
(2)   - copy pandocomatic-45.yaml
(1)   - copy test-img-2.png
Pandocomatic needed 1.3 seconds to convert 'pandocomatic-45' to 'out'.
```

If I `cd` into the directory and call pandocomatic from there it works, but that defeats the whole purpose of "converting a directory tree of documents".

Is this something that should or could be supported by pandocomatic? At least it would be really useful for the "converting a directory tree" feature.

This is not related to windows, as I executed the same test in a debian box and got the same result with latest pandoc (stable) and pandocomatic.

In the uploaded example I explain a bit more what I tested and a comparison with html mode.
[pandocomatic-45.zip](https://github.com/htdebeer/pandocomatic/files/1318046/pandocomatic-45.zip)

Thanks!!

**Comment by htdebeer** [comment]
To be honest, I never tried converting a directory tree to docx. I use it as a static-site generator and then it copies the images as well (by using the `recursive: true`) so the images do work as expected.

I agree with you that it would be nice if it should work with formats that integrate the images like docx (and pdf?) as well. I am not sure if this is a trivial addition though. If I run pandoc directly, like

```
pandoc --from markdown --to docx -i pandocomatic-45/pandocomatic-45.md -o out.docx
```

pandoc cannot find the images either.

One way to solve this issue is by using a filter that would rewrite the image URLs if they point to a path on the local file system. That should not be too hard to do with paru. 

Questions:

- should such a filter be included in the pandocomatic distribution or be a separate download?
- are there more things that you know of that might need to use this filter?
- should this filter be run automatically or should the user add it explicitly to a template?
- if it should run automatically, for which output formats should it be run?

**Comment by agusmba** [comment]
What if instead of rewriting the image URLs in a filter you simply changed the current directory before calling pandoc?
That's what I was doing in my bash scripts and it worked fine (the code was not pretty though).

Roughly anytime I get a markdown file with absolute or relative path:

* store current working dir for later retrieval
* get the path part of the markdown file and cd into it
* call pandoc
* cd back to the original directory.

I don't know if it'd be easy to do in ruby but to me it sounds easier than going the filter route?

This changed directory behaviour could be activated with a flag in order not to break backwards compatibility if needed.

And this approach would be compatible with directory tree invokation and with single file also.

**Comment by agusmba** [comment]
Regarding the filter approach, and to answer your questions (which are also valid for the change-dir approach) as far as I understand pandoc (which is not much, since I started playing with it a week ago)

Questions:

>* should such a filter be included in the pandocomatic distribution or be a separate download?

I'd prefer that it be included

>* are there more things that you know of that might need to use this filter?

I'm not experienced enough with the tools to know about other use cases that could benefit from this

>* should this filter be run automatically or should the user add it explicitly to a template?

I would say explicitly either via template or command-line option in order to activate it. Even though I think it would be harmless to have it always on, there could be set-ups that depend on the current behaviour (?)

>* if it should run automatically, for which output formats should it be run?

pandoc has so many different output formats that I think it would be easier letting the users choose whether to activate this feature or not. Maybe if the output formats had metadata indicating which ones include external files and which ones only reference them it would be easier to generalize it (pandoc feature request in the future?)

**Comment by htdebeer** [comment]
I will look into changing directories for the "convert directory tree" mode. It might be a lot more work to create than a filter, but I agree with you that is seems to be more consistent with running pandoc in that directory a user would be familiar with.

**Comment by agusmba** [comment]
A couple of pointers I just saw in case they help:

* [Several ways of changing directory](https://stackoverflow.com/questions/10147913/ruby-run-shell-command-in-a-specific-directory)
* [Playing with filenames and directories](http://rubyforadmins.com/files-and-directories)

**Comment by htdebeer** [comment]
Thanks for the pointer, but I am familiar with Ruby programming :-)

**Comment by agusmba** [comment]
So embarrasing...

**Comment by htdebeer** [comment]
Don't worry. Maybe it is of help for someone else who want to start hacking on pandocomatic.

There is no such thing as too much documentation!

**Comment by htdebeer** [comment]
I've just created a fix for this issue. You can download and install version 0.2.0.4 to try.

**Comment by agusmba** [comment]
Did a couple test and I can only say: **beautiful!**
Thanks @htdebeer ! I can probably ditch now my ~50 line launching script (told you it wasn't pretty) and use pandocomatic directly!
And I see that you made it work for directory and direct file invokation, nice!!

**Comment by htdebeer** [comment]
You're welcome and thank you for noticing this issue. It is only through feed-back from users that free software becomes awesome!

**Comment by iandol** [comment]
Hi i'm getting empty warnings when I try to process to docx files, is this related to this change?

```
👉  pandocomatic -b Pinna.md
(3) + converting Pinna.md 2 times:
(2)   - convert Pinna.md -> Pinna.docx
pandoc	--bibliography=pinna.json \
	--csl=/Users/ian/.pandoc/csl/cell.csl \
	--standalone \
	--filter=/usr/local/bin/pandoc-citeproc \
	--filter=/Users/ian/.pandoc/filters/removeHR \
	--filter=/Users/ian/.pandoc/filters/authorRemoveHash \
	--filter=/Users/ian/.pandoc/filters/prependInstitute \
	--from=markdown \
	--to=docx \
	--reference-doc=/Users/ian/.pandoc/templates/custom.docx \
	--dpi=300 \
	--output=/Users/ian/Desktop/Pinna_mmd/Pinna.docx
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
[warning] Could not fetch resource '': PandocResourceNotFound ""
```

There are 8 images in this document, all in the same directory as the md file. Sorry don't have time to look through the code...

**Comment by htdebeer** [comment]
That is possible, yes. Although the changes were small, the way pandoc is called changed fundamentally from the directory pandocomatic was invoked in to the directory where the file to convert lives.

Two questions:

1. despite the warnings, is the file being converted successfully?
2. can you create a minimal example where this issue do occur?

**Comment by iandol** [comment]
I’m travelling between Beijing, Shanghai, Moscow and Milan so no chance of
a test case, but docs seems to convert ok so no major worries 🤓

**Comment by htdebeer** [comment]
Safe travels!

I'll reopen this issue to investigate further, see if I can find a way to reproduce it myself.

**Comment by htdebeer** [comment]
As of yet, I have been unable to reproduce the issue. 

I have created a configuration file with two templates (`todocx` and `tohtml`):

```{.yaml}
templates:
    todocx:
        pandoc:
            from: markdown
            to: docx
            dpi: 300
            reference-doc: template.docx

    tohtml:
        pandoc:
            from: markdown
            to: html
```

I then converted a file with two images using both templates

``` {.bash}
pandocomatic -d data-dir/ -c docxtemplate.yaml --debug -i test.md
```

I did not get any warning when converting the file with both pandoc1 and pandoc2. (For the pandoc1 test, I changed the `todocx` template  to use option `reference-docx`, of course.) The other options in your example should not have any effect on converting a file with images to docx.

Possible explanations why I am unable to reproduce the issue are:

- You're using a different operating system; I am using Linux
- You're using a newer version of pandoc2; I was using pandoc-amd64-0b05222a
